docs/docs/vm-fundamentals/builtins.md

```diff
@@ -4,4 +4,229 @@ sidebar_position: 2
 
 # Builtins
 
-Explain each builtin here. How do they work and what do they accomplish.
+Regular Cairo instructions are like regular Legos, they work, but for complex things they can be slow. Builtins are special Lego pieces that snap together easily, like pre-built wheels or doors. This makes building things much faster. In a Cairo context, builtins help write clean and efficient Cairo programs, just like these special Lego pieces help build cool things faster.
+
+This section provides an overview of the builtins functions available in the Cairo Virtual Machine (Cairo VM).
+
+### What is builtins in cairo?
+
+Builtins are predefined optimized low-level execution units which are added to the Cairo CPU board to perform predefined computations which are expensive to perform in Cairo.
+
+Imagine the Cairo VM as a powerful computer. Regular Cairo instructions are like the basic building blocks the computer can understand to perform calculations, but for some complex tasks, these building blocks can be slow to put together. Builtins are like special hardware additions to the CPU that allow the computer to perform specific tasks much faster. Just like a computer with a powerful graphics card runs games smoother, builtins help write efficient Cairo programs.
+
+### Builtins and Cairo memory
+
+**Builtins act as special extensions to the Cairo VM's memory, enforcing specific rules on designated areas.** 
+
+Imagine the VM memory as a filing cabinet, builtins are like add-on modules that create special folders within the cabinet and each builtin gets its own folder and dictates the rules for what can be stored there. For instance, a "range-check" builtin might create a folder where all values must be between 0 and a very large number.
+
+**Communication between the CPU and builtins happens entirely through this memory.** 
+
+The CPU doesn't directly call builtins, instead, it treats the builtins memory folder like a special device to use a builtin. A function gets a pointer to the builtins folder as an argument, the function then reads or writes data to specific cells within the folder, following the rules set by the builtin. This method is similar to how external devices interact with a computer, they don't talk directly to the CPU but use the memory space for communication.
+
+**Adding builtins doesn't require modifying the CPU itself.** 
+
+Both the CPU and builtins share the same memory space, making them efficient tools. Additionally, each function using a builtin receives a pointer and is responsible for returning a pointer to the next unused cell within the builtin's memory area. This ensures proper management of space allocated for each builtin instance.
```

Comment:
Although the initial analogy with Legos is good, I don't think explaining the whole meaning with analogies is helpful rather than confusing. Bultins should be explained as what they are: special segments of the Cairo VM that behave a little different. The bring efficiency to the table by doing certain operations that with Cario set of instructions is extremely expensive at the cost of extra complexity when proving the program.